agent.go

```diff
@@ -231,6 +231,7 @@ func NewAgent(config *AgentConfig) (*Agent, error) {
 		startedCh:        startedCtx.Done(),
 		startedFn:        startedFn,
 		chanState:        make(chan ConnectionState, 1),
+		chanCandidate:    make(chan Candidate, 1),
```

Comment:
Do they need to be buffered? Unless absolutely necessary (E.g.: you're sending & receiving in the same routine; which I would also avoid but doesn't seem to be the case here) I would default to un-buffered channels.

Reply:
Dropped! I don't believe so, I think this is left over from old solutions. I also removed the buffering for `chanState`

Reply:
Cool still passes :)

@at-wat I would love your sign-off before I merge! I don't want to add any tech debt in the process. After I fix the STUNTURNConcurrency test I think we will be in a great spot.

Reply:
I'm still not very sure that it never cause deadlock.
For example, doesn't `agent.Close()` in candidate/state callback cause deadlock?

Reply:
```go
func TestCloseInConnectionStateCallback(t *testing.T) {
  report := test.CheckRoutines(t)
  defer report()

  lim := test.TimeOut(time.Second * 5)
  defer lim.Stop()

  disconnectedDuration := time.Second
  failedDuration := time.Second
  KeepaliveInterval := time.Duration(0)

  cfg := &AgentConfig{
    Urls:                []*URL{},
    NetworkTypes:        supportedNetworkTypes,
    DisconnectedTimeout: &disconnectedDuration,
    FailedTimeout:       &failedDuration,
    KeepaliveInterval:   &KeepaliveInterval,
    taskLoopInterval:    500 * time.Millisecond,
  }

  aAgent, err := NewAgent(cfg)
  if err != nil {
    t.Error(err)
  }

  bAgent, err := NewAgent(cfg)
  if err != nil {
    t.Error(err)
  }

  isClosed := make(chan interface{})
  err = aAgent.OnConnectionStateChange(func(c ConnectionState) {
    switch c {
    case ConnectionStateConnected:
      aAgent.Close()
    case ConnectionStateClosed:
      close(isClosed)
    }
  })
  if err != nil {
    t.Error(err)
  }

  connect(aAgent, bAgent)

  <-isClosed
  assert.NoError(t, bAgent.Close())
}
```
looks causing deadlock.

Reply:
Maybe it's better to run the callback asynchronously?

Reply:
It was called asynchronously, but calling each callback asynchronously made order of the callbacks random.
(e.g. nil candidate can be passed before the final candidate)

Reply:
Note that the buffered channel only delays the deadlock. If you get unlucky and the buffer is full, the channel operations will still become blocking and it will still deadlock.

Reply:
Making the close call unblocking may be a better solution. (E.g. signal the runloop to close by closing some done channel).

Reply:
On the other hand, I agree that it is better to be fully synchronous (non-buffered) if deadlock can be avoided.
On the current implementation, agent internal state may be updated during processing callback.
But, it requires large refactor to do this, it would be better to be done in the next PR.

Reply:
It would be better to ensure any other functions don't cause deadlock.